### PR TITLE
[14.0] [MIG] sale exception

### PR DESCRIPTION
**Pull request by sebastienbeau**

### CODE REVIEW

**Review by atchuthan — APPROVED** [review]
:+1:

File: sale_exception/data/sale_exception_data.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        <field name="active" eval="False" />
+    </record>
+    <record id="excep_no_stock" model="exception.rule">
+        <field name="name">Not Enough Virtual Stock</field>
+        <field name="description">Not Enough Virtual Stock</field>
+        <field name="sequence">50</field>
+        <field name="model">sale.order.line</field>
+        <field
+            name="code"
+        >if obj.product_id and sale_line.product_id.type == 'product' and obj.product_id.with_context(warehouse=obj.order_id.warehouse_id.id).virtual_available &lt; obj.product_uom_qty: failed=True</field>
```
Comment (Saran440) [inline sale_exception/data/sale_exception_data.xml:31]:
```suggestion
        >if obj.product_id and obj.product_id.type == 'product' and obj.product_id.with_context(warehouse=obj.order_id.warehouse_id.id).virtual_available &lt; obj.product_uom_qty: failed=True</field>
```

**Review by Saran440 — COMMENTED** [review]
Funtional Test
Error code exception.

**Comment by Saran440** [comment]
If user is not access admin can't click button confirm
Selection_003

Step to error
1. Create new user > acces rights Administration is false
Selection_004
2. Go to Sale Exception Rules > active No ZIP code on destination
3. Create Quotation and select partner is not zip code
4. Login new user > select Quotation > Confirm
5. It throw error access rights.

File: sale_exception/data/sale_exception_data.xml (hunk 1 above)
Comment (gurneyalex) [reply sale_exception/data/sale_exception_data.xml:31]: The check must be updated in 14.0 to use  the field virtual_quantity_at_date of the sale line to do the check. 

See : https://github.com/odoo/odoo/blob/1604f23afdb5c18d00147e25d0bc40cd0b57620c/addons/sale_stock/models/sale_order.py#L270

**Review by gurneyalex — CHANGES_REQUESTED** [review]

**Comment by tarbib** [comment]
> If user is not access admin can't click button confirm
> Selection_003
> 
> Step to error
> 
> 1. Create new user > acces rights Administration is false
>    Selection_004
> 2. Go to Sale Exception Rules > active No ZIP code on destination
> 3. Create Quotation and select partner is not zip code
> 4. Login new user > select Quotation > Confirm
> 5. It throw error access rights.

Hello @Saran440, I cannot reproduce this issue. Do you still have the problem ?

**Comment by rousseldenis** [comment]
@sebastienbeau Do you mind if we replace this by #1571 ?

**Comment by sebastienbeau** [comment]
@rousseldenis yes you can, thanks for continuing the work, I am too busy right now

**Pull request closed without merging**